Lazy implementations of takeRight and dropRight

szeiger opened the pull request:
- Add these methods to IterableOnceOps
- Implement them in Iterator (with the necessary amount of caching but
  no more than that)
- Move the existing strict implementations from IterableOps to
  StrictOptimizedIterableOps
- Add View.(TakeRight|DropRight) based on Iterator
- Move IndexedSeqView implementations of TakeRight, Drop and DropRight
  up to SeqView
- Add new overrides in IndexedSeqView

Fixes https://github.com/scala/bug/issues/11275

szeiger committed:
Lazy implementations of takeRight and dropRight

- Move the existing strict implementations from IterableOps to
  StrictOptimizedIterableOps
- Add View.(TakeRight|DropRight) with private Iterator-based
  implementations that cache as little and as late as possible
- Move IndexedSeqView implementations of TakeRight, Drop and DropRight
  up to SeqView
- Add new overrides in IndexedSeqView

Fixes https://github.com/scala/bug/issues/11275

odd commented:
@szeiger How does this PR relate to discussion around https://github.com/scala/collection-strawman/pull/96? My understanding was that the 2017-06-27 collections meeting (https://github.com/scala/collection-strawman/issues/118) decided that the right-leaning operation variants was not desirable for `Iterator`/`IterableOnce` as they required buffering; have this decision been reverted?

odd commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -1189,6 +1223,80 @@ object Iterator extends IterableFactory[Iterator] {
+    private[this] var pos: Int = 0
+    private[this] var buf: Array[AnyRef] = _
+    def init(): Unit = if(buf eq null) {
+      buf = new Array[AnyRef](buflen)
```
Does @Ichoran's https://github.com/scala/collection-strawman/pull/96#discussion_r120459949 apply here as well?

Ichoran replied:
Yes, it should apply!  Preallocating when you don't actually know that anywhere near that much space is needed isn't a good general-purpose strategy, even if it is more efficient in the case where the bound is chosen wisely.

szeiger commented:
Good point, I didn't even remember the earlier discussion around these operators. I don't know if we ever really considered the kind of as-lazy-as-possible implementations as opposed to the earlier ones that traversed the underlying Iterator on construction.

We already have a few other methods that have to do intermediate buffering similar to my `takeRight` and `dropRight` implementations, in particular `distinct` / `distinctBy`, `span` and `duplicate`.

If we don't want to add more of these to `Iterator` we could keep them private for use by the new Views only.

Ichoran commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -1189,6 +1223,80 @@ object Iterator extends IterableFactory[Iterator] {
+      len = 0
+      while(underlying.hasNext) {
+        buf(pos) = underlying.next().asInstanceOf[AnyRef]
+        pos = (pos + 1) % buflen
```
It's faster to do a jump than modulus arithmetic, i.e. something like `pos += 1; if (pos == buflen) pos = 0` (assuming `buflen` is sensible, which it had better be if an array was allocated using it).

Ichoran commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -1189,6 +1223,80 @@ object Iterator extends IterableFactory[Iterator] {
+        x
+      }
+    }
+    override def drop(n: Int): Iterator[A] = {
```
Why just `drop`?  Now that everything is in a ring buffer, `take`, `takeRight`, `slice`, etc. all have efficient implementations.  Is there something special about `drop` that makes it more likely to be used than the others?

NthPortal replied:
I think it's because `it.takeRight(n).drop(k)` is equivalent to `it.takeRight(n - k)`, and this iterator is already an implementation for `takeRight`. On the other hand,

- `it.takeRight(n).take(k)` is only equivalent to `it.takeRight(x)` if `k >= n` (in which case `x = n`)
- `it.takeRight(n).slice(a, b)` is only equivalent to `it.takeRight(x)` if `b > n` (in which case `x = b - a`)

However, `it.takeRight(n).takeRight(k)` *is* equivalent to `it.takeRight(x)`, where `x = n min k`, so `takeRight` should probably be overridden as well.

(Disclaimer: there is probably at least one off-by-one error in the above; use with caution)

szeiger replied:
`drop` was easy to implement. With too many other PRs to look after, I'll leave further optimizations to others.

Ichoran commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -1189,6 +1223,80 @@ object Iterator extends IterableFactory[Iterator] {
+    }
+  }
+
+  private final class DropRightIterator[A](private[this] var underlying: Iterator[A], buflen: Int) extends AbstractIterator[A] {
```
The logic of this is pretty similar to that of `TakeRightIterator` except for the `init()`.  Can you maybe create an abstract class, and pull the conditional out so you only hit the multiply-dispatched `init` once instead of on every call?

szeiger replied:
I don't see how. Initialization is supposed to be as lazy as possible so you have to check on every call. The use of the buffer is also quite different between both classes. There's not a lot of common code that could be factored out.

Ichoran replied:
Yeah, I guess it's not similar enough.  Never mind.

Ichoran commented on src/library/scala/collection/StrictOptimizedIterableOps.scala:
```diff
@@ -255,4 +255,35 @@ trait StrictOptimizedIterableOps[+A, +CC[_], +C]
+    */
+  override def takeRight(n: Int): C = {
+    val b = newSpecificBuilder
+    b.sizeHintBounded(n, toIterable)
```
Does this hint take into account that `n` may be way larger than the number of things available?

szeiger replied:
Yes, that's why it's bounded by `toIterable`

Ichoran commented on test/scalacheck/scala/collection/IteratorProperties.scala:
```diff
@@ -0,0 +1,41 @@
+  def check(f: (IterableOnceOps[Int, IterableOnce, IterableOnce[Int]], Int) => IterableOnce[Int]): Prop = forAll(Arbitrary.arbitrary[Seq[Int]], smallInteger) { (s: Seq[Int], n: Int) =>
+    val indexed = s.toIndexedSeq // IndexedSeqs and their Iterators have a knownSize
+    val simple = new SimpleIterable(s) // SimpleIterable and its Iterator don't
+    val stream = Stream.from(s) // Lazy
```
Maybe better to use `LazyList`?

Ichoran reviewed (changes requested):
As @odd pointed out, we probably shouldn't add these methods to `Iterator` itself given that we're trying to keep eager operations out of `Iterator` and we already don't have these.  If we wanted to make the functionality, we could expose them as a separate class inside the `Iterator` or `View` companion.  Otherwise they probably should be secret implementations that `View` uses.

szeiger commented:
Updated:
- Rebased
- Using `ArrayBuffer` to avoid allocating large arrays when `n` in `takeRight(n)` or `dropRight(n)` is much larger than the (unknown) collection size
- No `takeRight` and `dropRight` on `IterableOnce`
